### ISSUE TITLE
Structures, Mines, and Caves, etc. don't generate correctly.

### ISSUE THREAD

**Issue by BrainSlugs83** [BrainSlugs83/NewBiospheresMod#6]
Make the structure generation stuff work -- that is, populate the domes with towns, and the undergrounds with caves, rivers, and abandoned mine shafts and other naturally occurring terrain/features, etc. (this is an issue with the original mod as well).

**Comment by BrainSlugs83** [comment]
Also, it would be awesome if the lake and terrain generation (etc.) were handled by the biome (so no hard coded spherical lakes, etc.)

**Comment by TheLoneWolfling** [comment]
This is currently the single biggest issue with this mod - large chunks of the game are unavailable due to lack of spawners / **end portal** / villages etc.

Some of these can be worked around, but not all of it.

**Comment by BrainSlugs83** [comment]
Agreed on importance.  Can you give an example of the workarounds that you are referring to?

**Comment by TheLoneWolfling** [comment]
Sure. For example: You can work around the lack of spawners by building a standard hostile mob trap instead (or go to the nether). Not as efficient, but meh. And it's relatively easy to light up enough to make them fast. You can work around the lack of villages by turning zombie villagers. You can work around the lack of glowstone (if the nether isn't enabled) by killing witches (as they do spawn as hostile mobs, albeit rarely.) or trading. You can get music disks by killing creepers with skeletons. You can get name tags by fishing or trading. You can get saddles and melons by trading.

Much the same sort of workarounds that you end up doing in a superflat world.

Unfortunately, there are a number of things that plain cannot be worked around. For example:

* Blaze rods (if the nether isn't enabled)
* Wither skeletons / wither / beacons (if the nether isn't enabled)
* Lilypads
* Glowstone (although see above)
* Nether wart (if the nether isn't enabled)
* Getting to the end
* Silverfish
* Packed ice
* Misc. stone bricks. (Won't be a problem in 1.8)
* Mossy cobblestone - though it won't be a problem in 1.8
* Melons will become completely unavailable in 1.8
* Anything that spawns only in loot chests (if the nether isn't enabled, or if it doesn't spawn in the nether fortress loot chests)
* Horse armor won't be available in 1.8 unless the nether is enabled.

**Comment by BrainSlugs83** [comment]
Interesting.

Yes, hopefully we'll be able to do normal structure generation and such soon.

I'm confused about the "nether not enabled" comments, because -- I don't think there's a way to disable the nether (or was this added in 1.8? or am I missing something?) -- the "Nether" seemed to work for me when I tested it. -- I thought it was "The End" that was broken?

Another workaround for glowstone (or missing blocks) is to put them as part of the bridges, etc. (Glowstone makes for okay bridge rails).

**Comment by TheLoneWolfling** [comment]
It's only with a dedicated server: `allow-nether=false` in server properties.

It's something I tend to enable because the nether doesn't really fit the whole biospheres / skyblock thing. Especially not with the nether biospheres.

And I'm not sure how one would fit in lilypads / packed ice as bridge rails.

**Comment by trevordebrecht** [comment]
@BrainSlugs83 If you're ok with it, I'll start bringing over the changes from my fork that do actual terrain generation. Structures and caves are fully supported already.

**Comment by TheLoneWolfling** [comment]
@trevordebrecht Go ahead and make a pull request. Worse comes to worse, BrainSlugs says no.

Are the structures and caves configurable? Specifically, can cave generation be toggled on / off?

**Comment by BrainSlugs83** [comment]
@trevordebrecht, yes, please do!  -- Keep in mind, I'm still targeting 1.7.x, and I think there's been a bit of refactoring and changes to my branch since you originally forked.  (I plan to do future refactoring to try and target other frameworks as well. -- I will probably move the forge specific stuff into it's own package/namespace.)

The things that you've already fixed that I still need to fix, I am basing off your code where I can.

For example, the spawning fix I made was based on your work, but heavily adapted (some bugs came out of that that I'll need to fix, but, the idea was that it wouldn't be limited to just glass; also I left the WorldProvider registered, albeit immensely simplified -- simply to prevent spawning in the void, which seemed to still be a problem).
